Autocompletion in zsh conflicts with NVM's autocompletion

Hi mate,

If you remember we discussed the support for ZSH (#31) and you managed to get it working including the auto-completion (if I'm not mistaken). I think it did work specifically for sub-commands but I'm not sure.
Now I noticed that the auto-completion doesn't work for sub-commands in (it works just fine for the root commands). I have no idea what has broken the feature. Can you please help me to troubleshoot this?
My setup is (for the record, I'm setting it up on a clean machine):

**Alf version**: 92c38febbb80dfad27554612cf2c9e2edbebcf8b
**Bash version**: GNU bash, version 5.1.4(1)-release (x86_64-apple-darwin20.2.0)
**ZSH version**: zsh 5.8 (x86_64-apple-darwin20.0)
**OS version**: macOS 11.2.3 20D91

Thanks.

Thanks for the quick reply. So I went back and checked all the changes one by one (didn't know when exactly it broke) and found this in my `.zshrc`:
```bash
export NVM_DIR="$HOME/.nvm"
[ -s "$NVM_DIR/nvm.sh" ] && \. "$NVM_DIR/nvm.sh"  # This loads nvm
[ -s "$NVM_DIR/bash_completion" ] && \. "$NVM_DIR/bash_completion"  # This loads nvm bash_completion
```
This is after `source ~/bash_aliases` line and when I move it before that everything seems to work just fine. The problematic part is the bash completion script. I'm not a bash guru, you might find this interesting (`~/.nvm/bash_completion`, especially the last lines talking about bash vs zsh):
```bash
#!/usr/bin/env bash

# bash completion for Node Version Manager (NVM)

if ! command -v nvm &> /dev/null; then
  return
fi

__nvm_generate_completion() {
  declare current_word
  current_word="${COMP_WORDS[COMP_CWORD]}"
  # shellcheck disable=SC2207
  COMPREPLY=($(compgen -W "$1" -- "${current_word}"))
  return 0
}

__nvm_commands() {
  declare current_word
  declare command

  current_word="${COMP_WORDS[COMP_CWORD]}"

  COMMANDS='
    help install uninstall use run exec
    alias unalias reinstall-packages
    current list ls list-remote ls-remote
    install-latest-npm
    cache deactivate unload
    version version-remote which'

  if [ ${#COMP_WORDS[@]} == 4 ]; then

    command="${COMP_WORDS[COMP_CWORD - 2]}"
    case "${command}" in
      alias) __nvm_installed_nodes ;;
    esac

  else

    case "${current_word}" in
      -*) __nvm_options ;;
      *) __nvm_generate_completion "${COMMANDS}" ;;
    esac

  fi
}

__nvm_options() {
  OPTIONS=''
  __nvm_generate_completion "${OPTIONS}"
}

__nvm_installed_nodes() {
  __nvm_generate_completion "$(nvm_ls) $(__nvm_aliases)"
}

__nvm_aliases() {
  declare aliases
  aliases=""
  if [ -d "${NVM_DIR}/alias" ]; then
    aliases="$(cd "${NVM_DIR}/alias" && command find "${PWD}" -type f | command sed "s:${PWD}/::")"
  fi
  echo "${aliases} node stable unstable iojs"
}

__nvm_alias() {
  __nvm_generate_completion "$(__nvm_aliases)"
}

__nvm() {
  declare previous_word
  previous_word="${COMP_WORDS[COMP_CWORD - 1]}"

  case "${previous_word}" in
    use | run | exec | ls | list | uninstall) __nvm_installed_nodes ;;
    alias | unalias) __nvm_alias ;;
    *) __nvm_commands ;;
  esac

  return 0
}

# complete is a bash builtin, but recent versions of ZSH come with a function
# called bashcompinit that will create a complete in ZSH. If the user is in
# ZSH, load and run bashcompinit before calling the complete function.
if [[ -n ${ZSH_VERSION-} ]]; then
  autoload -U +X bashcompinit && bashcompinit
  autoload -U +X compinit && if [[ ${ZSH_DISABLE_COMPFIX-} = true ]]; then
    compinit -u
  else
    compinit
  fi
fi

complete -o default -F __nvm nvm
```
Do you think the comment above is relevant here?

--- a/README.md
+++ b/README.md
@@ -157,6 +157,20 @@ should add this line to your startup script:
 source ~/.bash_aliases
 ```
 
+### ZSH Users
+
+Oh-My-Zsh users may skip this step.
+
+If you would like to have autocompletion for alf's sub-aliases and you are using
+zsh, you should enable completion by adding this to your `~/.zshrc` (if is it not
+already there) before sourcing `.bash_aliases`:
+
+```bash
+# Load completion functions
+autoload -Uz +X compinit && compinit
+autoload -Uz +X bashcompinit && bashcompinit
+```
+
 
 Related Projects
 --------------------------------------------------

--- a/alf
+++ b/alf
@@ -390,25 +390,13 @@ generate_completions() {
   find_config
 
   # shellcheck disable=SC2016
-  echo '
-# Enable bash completions in zsh
-if [[ -n ${ZSH_VERSION-} ]]; then
-  if ! command -v compinit > /dev/null; then
-    autoload -U +X compinit && if [[ ${ZSH_DISABLE_COMPFIX-} = true ]]; then
-      compinit -u
-    else
-      compinit
-    fi
-  fi
-  autoload -U +X bashcompinit && bashcompinit
-fi
-'
   echo '# Completions'
+  echo 'if command -v complete &> /dev/null ; then'
 
   while IFS= read -r line || [ -n "$line" ]; do
     if [[ $line =~ $ali1_regex ]]; then
       ali="${BASH_REMATCH[1]}"
-      [[ -n $comp ]] && echo "complete -W $comp"
+      [[ -n $comp ]] && echo "  complete -W $comp"
       comps=""
       comp=""
     elif [[ $line =~ $ali2_regex ]]; then
@@ -422,8 +410,10 @@ fi
   done < "$config_file"
   
   if [[ -n $comp ]] && [[ -n $comps ]] ; then
-    echo "complete -W $comp"
+    echo "  complete -W $comp"
   fi
+  
+  echo 'fi'
 }
 
 # :src/lib/generate_config.sh

--- a/src/lib/generate_completion.sh
+++ b/src/lib/generate_completion.sh
@@ -4,25 +4,13 @@ generate_completions() {
   find_config
 
   # shellcheck disable=SC2016
-  echo '
-# Enable bash completions in zsh
-if [[ -n ${ZSH_VERSION-} ]]; then
-  if ! command -v compinit > /dev/null; then
-    autoload -U +X compinit && if [[ ${ZSH_DISABLE_COMPFIX-} = true ]]; then
-      compinit -u
-    else
-      compinit
-    fi
-  fi
-  autoload -U +X bashcompinit && bashcompinit
-fi
-'
   echo '# Completions'
+  echo 'if command -v complete &> /dev/null ; then'
 
   while IFS= read -r line || [ -n "$line" ]; do
     if [[ $line =~ $ali1_regex ]]; then
       ali="${BASH_REMATCH[1]}"
-      [[ -n $comp ]] && echo "complete -W $comp"
+      [[ -n $comp ]] && echo "  complete -W $comp"
       comps=""
       comp=""
     elif [[ $line =~ $ali2_regex ]]; then
@@ -36,7 +24,9 @@ fi
   done < "$config_file"
   
   if [[ -n $comp ]] && [[ -n $comps ]] ; then
-    echo "complete -W $comp"
+    echo "  complete -W $comp"
   fi
+  
+  echo 'fi'
 }
 

--- a/test/fixtures/generate/aliases.txt
+++ b/test/fixtures/generate/aliases.txt
@@ -86,22 +86,11 @@ dns() {
   fi
 }
 
-
-# Enable bash completions in zsh
-if [[ -n ${ZSH_VERSION-} ]]; then
-  if ! command -v compinit > /dev/null; then
-    autoload -U +X compinit && if [[ ${ZSH_DISABLE_COMPFIX-} = true ]]; then
-      compinit -u
-    else
-      compinit
-    fi
-  fi
-  autoload -U +X bashcompinit && bashcompinit
-fi
-
 # Completions
-complete -W "s l p" g
-complete -W "help" abc
-complete -W "again" say
-complete -W "ls deploy upd" dc
-complete -W "check flush" dns
+if command -v complete &> /dev/null ; then
+  complete -W "s l p" g
+  complete -W "help" abc
+  complete -W "again" say
+  complete -W "ls deploy upd" dc
+  complete -W "check flush" dns
+fi

--- a/test/fixtures/generate/approvals/alf_generate
+++ b/test/fixtures/generate/approvals/alf_generate
@@ -86,22 +86,11 @@ dns() {
   fi
 }
 
-
-# Enable bash completions in zsh
-if [[ -n ${ZSH_VERSION-} ]]; then
-  if ! command -v compinit > /dev/null; then
-    autoload -U +X compinit && if [[ ${ZSH_DISABLE_COMPFIX-} = true ]]; then
-      compinit -u
-    else
-      compinit
-    fi
-  fi
-  autoload -U +X bashcompinit && bashcompinit
-fi
-
 # Completions
-complete -W "s l p" g
-complete -W "help" abc
-complete -W "again" say
-complete -W "ls deploy upd" dc
-complete -W "check flush" dns
+if command -v complete &> /dev/null ; then
+  complete -W "s l p" g
+  complete -W "help" abc
+  complete -W "again" say
+  complete -W "ls deploy upd" dc
+  complete -W "check flush" dns
+fi

